### PR TITLE
Add FHIRTaskExpireUtil to fetch and mark expired tasks

### LINKED ISSUE #1703: [Quest] Write a service to mark Tasks as expired.
**Enhancement context**
- We currently, only show the `overdue` tasks. However, some tasks get expired after a certain period of them being `overdue`
- These expired tasks can't be performed anymore. We will be tracking the expired tasks by updating the `Task.status` to `cancelled`. 
- We will be tracking the expiry date from the `Task.restriction.period.end`. If the end date is =< `today()` then the task is expired. 

**Implementation**
- [x] Write a service to mark tasks as expired. The timing of the service should be at least after every 1 hour. 
   - [x] The timing should be configurable on the `application_config`.  In case it's not added then the default is set to 1 hour. 
- [x] The `Task.restriction.period.end` should be used to determine the task expiry. 
   - [x] If the `Task.restriction.period.end` date =< `today()` then update the `Task.status` to `cancelled` 

**Acceptance Criteria**
- [x] All the expired tasks should have a status `cancelled`. 
- [x] The service to update the statuses should run periodically.

### PR DESCRIPTION
**Pull request by ekigamba**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #1703

**Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [x] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the fhircore app to verify my change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireUtil.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import org.smartregister.fhircore.engine.util.extension.isPastExpiry
+import org.smartregister.fhircore.engine.util.extension.toCoding
+import timber.log.Timber
+
+class FhirTaskExpireUtil(val appContext: Context) {
+
+
+  internal fun getFhirEngine() : FhirEngine {
+    return FhirEngineProvider.getInstance(appContext)
+  }
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireUtil.kt:33]: I think we can inject the `FhirEngine` dependency instead.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireUtil.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+
+  suspend fun fetchOverdueTasks() : List<Task> {
+    Timber.i("Start fetching overdue tasks")
+
+    val fhirEngine = getFhirEngine()
+
+    // TODO: Limit & page the results for better performance
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireUtil.kt:41]: Will using the `paging` library solve the issue with pagination?

**Comment by ellykits** [comment]
I think we need to move this implementation here [FhirTaskPlanWorker](https://github.com/opensrp/fhircore/blob/d2ab5130917cf1df2aa1ea189133bff380827663/android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskPlanWorker.kt#L35)

This is a job that can be scheduled to update the statuses of all tasks.

CC: @dubdabasoduba @maimoonak

**Comment by ekigamba** [comment]
@ellykits I see that the implementation provided runs the job every 12 hours while the requirements for this job talked about every 1 hour [here](https://github.com/opensrp/fhircore/issues/1703#issue-1389996502). Otherwise, the `FhirTaskExpireUtil` is not a job and only contains the functionality so we can call it from the `FhirTaskPlanWorker` job

**Comment by dubdabasoduba** [comment]
> @ellykits I see that the implementation provided runs the job every 12 hours while the requirements for this job talked about every 1 hour [here](https://github.com/opensrp/fhircore/issues/1703#issue-1389996502). Otherwise, the `FhirTaskExpireUtil` is not a job and only contains the functionality so we can call it from the `FhirTaskPlanWorker` job

@ellykits @ekigamba Can we write the Worker to allow a custom run period set up? i.e we make it possible for each job to define its own execution period.

**Comment by ellykits** [comment]
Sure I meantioned this in another PR by @ahsanbhatti49

**Comment by f-odhiambo** [comment]
@dubdabasoduba Ahsan is working on it on background Tasks so these should be added as Jobs to it

**Comment by ahsanbhatti49** [comment]
> > @ellykits I see that the implementation provided runs the job every 12 hours while the requirements for this job talked about every 1 hour [here](https://github.com/opensrp/fhircore/issues/1703#issue-1389996502). Otherwise, the `FhirTaskExpireUtil` is not a job and only contains the functionality so we can call it from the `FhirTaskPlanWorker` job
> 
> @ellykits @ekigamba Can we write the Worker to allow a custom run period set up? i.e we make it possible for each job to define its own execution period.

We can make a periodic worker that can run after a specific time interval. and define this interval in the config. 
Interval can be specified by Sec,Minutes,hours etc. Lets say if we want to run a job after every 24 hours we can do that. But there is no direct way of specifying that a job should run daily at specific time. for example: daily at 5pm. etc.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireJob.kt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import dagger.hilt.android.EntryPointAccessors
+import java.util.concurrent.TimeUnit
+import org.smartregister.fhircore.engine.di.FhirTaskExpireEntryPoint
+import org.smartregister.fhircore.engine.util.SharedPreferencesHelper
+
+class FhirTaskExpireJob constructor(val context: Context, workerParams: WorkerParameters) :
+  CoroutineWorker(context, workerParams) {
+
+  var fhirEngine: FhirEngine = getFhirEngine(context)
+  var fhirTaskExpireUtil: FhirTaskExpireUtil = getFhirTaskExpireUtil(context)
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireJob.kt:35]: It's possible to inject dependencies to `CoroutineWorker`. See [AppSyncWorker](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/sync/AppSyncWorker.kt) for example.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/di/FhirTaskExpireEntryPoint.kt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+/** Created by Ephraim Kigamba - nek.eam@gmail.com on 05-12-2022. */
+@InstallIn(SingletonComponent::class)
+@EntryPoint
+interface FhirTaskExpireEntryPoint {
+
+  fun fhirEngine(): FhirEngine
+
+  fun fhirTaskExpireUtil(): FhirTaskExpireUtil
+}
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/di/FhirTaskExpireEntryPoint.kt:33]: Was this required for dependency injection? Dependencies can be injected to a subclass  of `CoroutineWorker` as constructor parameters.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireJob.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      if (currVersion != version) {
+        existingWorkPolicy = ExistingPeriodicWorkPolicy.REPLACE
+        sharedPreferencesHelper.write(FHIR_TASK_EXPIRE_JOB_VERSION, version)
+      }
+
+      val periodicWorkRequest =
+        PeriodicWorkRequestBuilder<FhirTaskExpireJob>(
+            durationInMins,
+            TimeUnit.MINUTES,
+            5,
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireJob.kt:74]: Do we need a TODO to configure this in future?

**Review by ellykits — CHANGES_REQUESTED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireJob.kt (hunk 3 above)
Comment (ekigamba) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireJob.kt:35]: I believe so but I spent a lot of time trying it but couldn't get it to work. 

You can possibly help to get this working. The current injection works almost similar and depends on Hilt

File: android/engine/src/main/java/org/smartregister/fhircore/engine/di/FhirTaskExpireEntryPoint.kt (hunk 4 above)
Comment (ekigamba) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/di/FhirTaskExpireEntryPoint.kt:33]: I believe so but I spent a lot of time trying it but couldn't get it to work. 

You can possibly help to get this working. The current injection works almost similar and depends on Hilt

File: android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireJob.kt (hunk 5 above)
Comment (ekigamba) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskExpireJob.kt:74]: Nice catch. I have updated the code to make the `durationInMins` configurable. The magic number `5` is for the `flexTimeInterval` which I don't expect to change and might be difficult to explain it's impact as a configurable number

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/app/ApplicationConfiguration.kt [hunk 6]
```diff
@@ -36,5 +36,5 @@ data class ApplicationConfiguration(
   val deviceToDeviceSync: DeviceToDeviceSyncConfig? = null,
   val snackBarTheme: SnackBarThemeConfig = SnackBarThemeConfig(),
   val reportRepeatTime: String = "",
-  val taskExpireJobRepeatIntervalConfig: Long = 60
+  val taskExpireJobRepeatInterval: Long = 60
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/app/ApplicationConfiguration.kt:39]: is this in minutes? I think we should add that to the name so it's clear
```suggestion
  val taskExpireJobRepeatIntervalMinutes: Long = 60
```

File: android/quest/src/main/assets/configs/app/application_config.json [hunk 7]
```diff
@@ -6,6 +6,7 @@
   "registerDate": "2020-10-27",
   "reportRepeatTime": "23:22",
   "remoteSyncPageSize": 100,
+  "taskExpireJobRepeatInterval": "15",
```
Comment (pld) [inline android/quest/src/main/assets/configs/app/application_config.json:9]:
```suggestion
  "taskExpireJobRepeatIntervalMinutes": "15",
```

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/main/AppMainViewModel.kt [hunk 8]
```diff
@@ -322,10 +322,10 @@ constructor(
       applicationConfiguration.registerDate
     )
 
-    FhirTaskExpireJob.schedule(
+    FhirTaskExpireWorker.schedule(
       workManager,
       sharedPreferencesHelper,
-      applicationConfiguration.taskExpireJobRepeatIntervalConfig
+      applicationConfiguration.taskExpireJobRepeatInterval
```
Comment (pld) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/main/AppMainViewModel.kt:328]:
```suggestion
      applicationConfiguration.taskExpireJobRepeatIntervalMinutes
```

**Review by pld — CHANGES_REQUESTED** [review]

**Review by ellykits — DISMISSED** [review]

**Review by pld — APPROVED** [review]

**Pull request merged**